http-clients/src/main/java/com/palantir/remoting3/clients/UserAgents.java

```diff
@@ -73,7 +83,78 @@ private static String formatSingleAgent(UserAgent.Agent agent, Map<String, Strin
         return formatted.toString();
     }
 
-    static boolean isValidLibraryName(String serviceName) {
+
+    /**
+     * Parses the given string into a {@link UserAgent} or throws an {@link IllegalArgumentException} if the string does
+     * not conform to the user agent syntax restrictions.
+     */
+    public static UserAgent parse(String userAgent) {
+        return parseInternal(userAgent, false /* strict */);
+    }
+
+    /** Like {@link #parse}, but returns only the syntactically correct agent parts (, possibly zero). */
+    public static UserAgent tryParse(String userAgent) {
+        return parseInternal(userAgent, true /* lenient*/);
+    }
+
+    private static UserAgent parseInternal(String userAgent, boolean lenient) {
+        ImmutableUserAgent.Builder builder = ImmutableUserAgent.builder();
+        List<String> parts = COMMA_SPLITTER.splitToList(userAgent);
+        checkArgument(!parts.isEmpty(), "Empty user agents are not allowed");
+
+        // Primary agent with optional nodeId
+        Map<String, String> primaryComments = Maps.newHashMap();
+        Optional<UserAgent.Agent> primaryAgent = tryParseSingleAgent(parts.get(0), primaryComments);
+        if (primaryAgent.isPresent()) {
+            builder.primary(primaryAgent.get());
+        } else {
+            if (lenient) {
+                builder.primary(UserAgent.Agent.of("unknown", UserAgent.Agent.DEFAULT_VERSION));
+            } else {
+                throw new IllegalArgumentException("Agent string was malformed: " + parts.get(0));
+            }
+        }
+        if (primaryComments.containsKey("nodeId")) {
+            builder.nodeId(primaryComments.get("nodeId"));
+        }
+
+        // Informational agents
+        for (String part : parts.subList(1, parts.size())) {
+            Optional<UserAgent.Agent> informationlAgent = tryParseSingleAgent(part, Maps.newHashMap());
+            if (informationlAgent.isPresent()) {
+                builder.addInformational(informationlAgent.get());
+            } else {
+                if (lenient) {
+                    // Ignore malformed agent
+                    log.warn("Cannot parse malformed user agent string", UnsafeArg.of("agent", part));
```

Comment:
why is this case different than the primary agent case? could this also be `unknown/0.0.0`?

Reply:
because its presence is optional anyway